### PR TITLE
WIP - py3 compatibility: fix of remaining issues and enabling Python 3 FAF package

### PR DESCRIPTION
**Pull request by ari3s**
In Python 2, the `str` type can be used for both text and binary data. It does not work in Python 3 where must be distinguished which APIs take text and which take binary. There must be also considered that Python 2 text APIs can work with `unicode` and APIs that work with binary data, must work with the `bytes` type from Python 3.

Signed-off-by: Jan Beran <jberan@redhat.com>

### CODE REVIEW

**Comment by ari3s** [comment]
retest this please

**Comment by ari3s** [comment]
retest this please

File: faf.spec.in [hunk 1]
```diff
@@ -21,23 +27,39 @@ Source0: https://github.com/abrt/%{name}/archive/%{version}/%{name}-%{version}.t
 
 BuildArch: noarch
 
+%if 0%{?python3}
```
[the remark this answers is not in the document]
Comment (ari3s) [reply faf.spec.in]: Done, thank you.

File: faf.spec.in [hunk 2]
```diff
@@ -46,17 +68,34 @@ BuildRequires: python2-six
...
+BuildRequires: python3-alembic
+BuildRequires: python3-setuptools
+BuildRequires: python3-unittest2
+BuildRequires: python3-bugzilla >= 2.0
+BuildRequires: python3-psycopg2 >= 2.5
+BuildRequires: python3-testing.postgresql < 1.2
+BuildRequires: python3-rpm
+BuildRequires: python3-sqlalchemy >= 0.8
+BuildRequires: python3-koji
+BuildRequires: python3-zeep
```
[the remark this answers is not in the document]
Comment (ari3s) [reply faf.spec.in]: I have updated the spec file.

File: faf.spec.in [hunk 3]
```diff
@@ -21,23 +27,39 @@ Source0: https://github.com/abrt/%{name}/archive/%{version}/%{name}-%{version}.t
...
 
 Requires(pre): shadow-utils
 
 Requires: postgresql
+
+%if %{with python3}
+Requires: python3
+Requires: python3-setuptools
+Requires: python3-psycopg2 >= 2.5
+Requires: python3-sqlalchemy >= 0.8
```
Comment (xsuchy) [inline faf.spec.in]: python3-psycopg2 < 2.5 and python3-sqlalchemy < 0.8 does not exist. You can remove the version from the requires.

File: faf.spec.in [hunk 4]
```diff
@@ -46,17 +68,34 @@ BuildRequires: python2-six
...
+BuildRequires: createrepo_c
+BuildRequires: python3
+BuildRequires: python3-alembic
+BuildRequires: python3-setuptools
+BuildRequires: python3-unittest2
+BuildRequires: python3-bugzilla >= 2.0
+BuildRequires: python3-psycopg2 >= 2.5
+BuildRequires: python3-testing.postgresql < 1.2
+BuildRequires: python3-rpm
+BuildRequires: python3-sqlalchemy >= 0.8
```
Comment (xsuchy) [inline faf.spec.in]: And same here for all >= operands.

File: faf.spec.in [hunk 5]
```diff
@@ -65,23 +104,45 @@ BuildRequires: python-psycopg2 >= 2.5
...
 %endif
-BuildRequires: %{satyr_dep}
-BuildRequires: python2-koji
-BuildRequires: fedmsg
-%if 0%{?fedora}
-BuildRequires: python3-zeep
-BuildRequires: python3-pylint
+
+%if %{with python3}
+BuildRequires: python-zeep
```
Comment (xsuchy) [inline faf.spec.in]: I believe this should be python3-zeep - for all instance in the spec.

File: faf.spec.in [hunk 6]
```diff
@@ -65,23 +104,45 @@ BuildRequires: python-psycopg2 >= 2.5
...
+
 # webui
+%if %{with python3}
+BuildRequires: python3-flask >= 0.10
+BuildRequires: python3-flask-wtf
+BuildRequires: python3-flask-openid
+BuildRequires: python3-openid-teams
+BuildRequires: python3-flask-rstpages
+BuildRequires: python3-flask-sqlalchemy
+BuildRequires: python3-jinja2 >= 2.7
```
Comment (xsuchy) [inline faf.spec.in]: Again all >= * can be omitted in this block.

File: faf.spec.in [hunk 7]
```diff
@@ -108,6 +165,20 @@ development.
...
 Requires: httpd
+%if %{with python3}
+Requires: python3-mod_wsgi
+Requires: python3-flask >= 0.10
+Requires: python3-flask-wtf
+Requires: python3-flask-rstpages
+Requires: python3-flask-sqlalchemy
+Requires: python3-flask-openid
+Requires: python3-openid-teams
+Requires: python3-jinja2 >= 2.7
```
Comment (xsuchy) [inline faf.spec.in]: And here.

File: faf.spec.in [hunk 8]
```diff
@@ -459,7 +546,11 @@ A plugin for %{name} implementing checking of repositories
 %package bugtracker-bugzilla
 Summary: %{name}'s bugzilla support
 Requires: %{name} = %{faf_version}
+%if %{with python3}
+Requires: python3-bugzilla >= 2.0
```
Comment (xsuchy) [inline faf.spec.in]: And here.

File: faf.spec.in [hunk 9]
```diff
@@ -509,24 +600,32 @@ Requires: faf = %{faf_version}
...
+%endif
 
 %description blueprint-celery-tasks
 A plugin for %{name} implementing Celery tasks blueprint plugin.
 
 %package migrations
 Summary: %{name}'s database migrations
 Requires: %{name} = %{faf_version}
+%if %{with python3}
+Requires: python3-alembic >= 0.5
```
Comment (xsuchy) [inline faf.spec.in]: and here

File: faf.spec.in (hunk 7 above)
Comment (ari3s) [reply faf.spec.in]: @xsuchy @marusak Thanks for the review. I have updated the spec file according to the feedback and our conversation today morning.

File: faf.spec.in (hunk 8 above)
Comment (ari3s) [reply faf.spec.in]: I keep the version condition here because python3-bugzilla < 2.0 was built in Fedora.

File: tests/test_rpm_metadata.py [hunk 10]
```diff
@@ -97,7 +97,7 @@ def setUp(self):
                         os.path.join(self.tmpdir, os.path.basename(self.rpm)))
 
         proc = popen("createrepo", self.tmpdir)
-        self.assertIn("Workers Finished", proc.stdout)
+        self.assertIn(b"Workers Finished", proc.stdout)
```
Comment (marusak) [inline tests/test_rpm_metadata.py]: This test fails. With createrepo_c the output of the command createrepo differes. With python 3 it should test Pool finished and not Workers Finished. However we need to test Workers Finished for python2, so we need to check if one of those is present.

Comment (ari3s) [reply tests/test_rpm_metadata.py]: @marusak Thank you, I have pushed the commit.

File: faf.spec.in [hunk 11]
```diff
@@ -93,11 +154,7 @@ BuildRequires: python-munch
 BuildRequires: python-dateutil
 BuildRequires: python2-ratelimitingfilter
 BuildRequires: python-simplejson
-BuildRequires: xstatic-patternfly-common
-BuildRequires: nodejs-typeahead.js
-BuildRequires: nodejs-moment
-BuildRequires: js-jquery
-BuildRequires: xstatic-datatables-common
```
Comment (marusak) [inline faf.spec.in]: Why were these packages removed?

File: faf.spec.in [hunk 12]
```diff
@@ -120,11 +191,7 @@ Requires: python-munch
 Requires: python-dateutil
 Requires: python2-ratelimitingfilter
 Requires: python-simplejson
-Requires: xstatic-patternfly-common
-Requires: nodejs-typeahead.js
-Requires: nodejs-moment
-Requires: js-jquery
-Requires: xstatic-datatables-common
```
Comment (marusak) [inline faf.spec.in]: Why were these packages removed?

File: faf.spec.in (hunk 11 above)
Comment (ari3s) [reply faf.spec.in]: It is my mistake, I have returned the lines back.

File: faf.spec.in (hunk 12 above)
Comment (ari3s) [reply faf.spec.in]: It is my mistake, I have returned the lines back.

File: src/pyfaf/utils/decorators.py [hunk 13]
```diff
@@ -66,7 +66,7 @@ def f_retry(*args, **kwargs):
                 time.sleep(mdelay)
                 mdelay *= backoff  # make future wait longer
 
-            raise ex  # out of tries
+            raise exc_value  # out of tries
```
[the remark this answers is not in the document]
Comment (marusak) [reply src/pyfaf/utils/decorators.py]: https://stackoverflow.com/questions/44258903/scope-of-caught-exception-instance-in-python-2-and-3
Found answer to this. It really changed in python3. Following script can demonstrate this behavior between python2 and python3.


    for i in range(5):                                                              
        try:                                                                        
            raise Exception(i)                                                      
        except Exception as ex:                                                     
            value = i                                                               
                                                                                  
    print(value)                                                                    
    print(i)                                                                        
    print(ex)

**Pull request closed without merging**